File: java/linkedlist/SortList.java
package linkedlist;

public class SortList {
    /* 148
     * 感觉思路并不难 就是归并排序merge sort 重点在于细节
     * 首先是寻找中点 由于不仅仅是要找中间点 而且还要linkedlist分成两半
     * 不然的话两个子list就会重复 所以感觉之前找中间点的代码会出现问题 加入一个prev之后 prev得不到更新 如果从dummy开始
     * 故此 先判断是否head head.next为null 然后可以让slow fast直接从head 和head.next开始 注意fast != null && fast.next == null
     * 意味着偶数size slow要再前进一格
     * merge的细节在于剩下的不用while再加入 直接copy listnode就可以了
     */
    public ListNode sortList(ListNode head) {
        return mergeSort(head);
    }
    private ListNode mergeSort(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode slow = head;
        ListNode fast = head.next;
        ListNode prev = head;
        while(fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        if(fast != null) {
            prev = slow;
            slow = slow.next;
        }
        prev.next = null;
        ListNode left = mergeSort(head);
        ListNode right = mergeSort(slow);
        return merge(left, right);
    }
    private ListNode merge(ListNode a, ListNode b) {
        ListNode dummy = new ListNode();
        ListNode cur = dummy;
        while(a != null && b != null) {
            if(a.val < b.val) {
                cur.next = a;
                cur = cur.next;
                a = a.next;
            }else{
                cur.next = b;
                cur = cur.next;
                b = b.next;
            }
        }
        if(a != null) {
            cur.next = a;
        }
        if(b != null) {
            cur.next = b;
        }
        return dummy.next;
    }
}

File: c++/linkedlist/sortList.cpp
#include <include/listnode.h>
/**
 * 基本思路是用快慢指针找到中点 然后merge sort
 * merge的代码和merge sort list的是一样的 cur = cur->next这一步要等到a = a->next 不然会被覆盖掉
 * 另外快慢指针可以用这套模版 两者都是从head开始 经过推算slow在中点 偶数list会在左边中点 
*/
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        return mergeSort(head);
    }
    ListNode* mergeSort(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode *slow = head, *fast = head;
        while(fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        fast = slow->next;
        slow->next = nullptr;
        return merge(mergeSort(head), mergeSort(fast));
    }

    ListNode* merge(ListNode* a, ListNode* b) {
        ListNode *dummy = new ListNode(0), *cur = dummy;
        while(a && b) {
            if(a->val <= b->val) {
                cur->next = a;
                a = a->next;
                cur = cur->next;
            }else{
                cur->next = b;
                b = b->next;
                cur = cur->next;
            }
        }
        if(a) cur->next = a;
        else if(b) cur->next = b;
        return dummy->next;
    }
};